clang/test/SemaSYCL/unique_stable_name.cpp

```diff
@@ -0,0 +1,216 @@
+
+#define MACRO12()                \
+  auto l12 = []() { return 1; }; \
+  constexpr const char *l1_output = __builtin_unique_stable_name(l12); // #USN_l12
+
+int main() {
+
+  // kernel1 - expect no error
+  // Evaluate the builtin in a constant context, then call the kernel with
+  // that same type
+  // Tests that passing the lambda to the unique stable name builtin and
+  // then using the lambda in a way that does not  contribute to the kernel
+  // name does not cause an error because the  stable name is not invalidated
+  // in this situation.
+  auto l1 = []() {};
+  constexpr const char *l1_output = __builtin_unique_stable_name(l1);
+  kernel_single_task<class kernel1>(
+      [=]() { l1(); }); // Used in the kernel, but not the kernel name itself
+
+  // kernel2 - expect error
+  // Call to the builtin on a lambda, then using the lambda for a kernel name
+  // Test that passing the lambda to the USN builtin and then using the same
+  // lambda in the naming of a kernel causes a diagnostic on the kernel use due
+  // to the change in results to the stable name.
+  auto l2 = []() { return 1; };
+  constexpr const char *l2_output = __builtin_unique_stable_name(l2); // #USN_L2
+  // expected-error@#kernelSingleTask{{kernel instantiation changes the result of an evaluated '__builtin_unique_stable_name'}}
+  // expected-note@#USN_L2{{'__builtin_unique_stable_name' evaluated here}}
+  // expected-note@+1{{in instantiation of function template specialization}}
+  kernel_single_task<class kernel2>(l2); // Used in the kernel name after builtin
+
+  // kernel7 - expect error
+  // Same as kernel6, except make l6 part of naming kernel
+  // Test that passing a lambda to the unique stable name builtin and then
+  // passing a second lambda to the kernel throws an error because the first
+  // lambda is included in the signature of the second lambda, hence attempting
+  // to change the mangling of the kernel
+  auto l7 = []() { return 1; };
+  auto l8 = [](decltype(l7) *derp = nullptr) { return 2; };
+  constexpr const char *l7_output = __builtin_unique_stable_name(l7); // #USN_l7
+  // expected-error@#kernelSingleTask{{kernel instantiation changes the result of an evaluated '__builtin_unique_stable_name'}}
+  // expected-note@#USN_l7{{'__builtin_unique_stable_name' evaluated here}}
+  // expected-note@#kernel7call{{in instantiation of function template specialization}}
+  kernel_single_task<class kernel7>(l8); // #kernel7call
+
+  // kernel8 and kernel9 - expect error
+  // Make two lambdas
+  // Call the builtin on the second of the two lambdas (in a constexpr context)
+  // Within a constexpr if, pass the first lambda to a kernel in the true branch,
+  // pass the second lambda to a kernel in the false branch
+  // Test that passing a lambda to the unique stable name builtin and passing it
+  // to a kernel called with an if constexpr branch causes a diagnostic on the
+  // kernel9 use due to the change in the results to the stable name
```

Comment:
```suggestion
  // Tests that passing a lambda to the unique stable name builtin and passing it
  // to a kernel called with an if constexpr branch causes a diagnostic on the
  // kernel9 use due to the change in the results to the stable name. This happens
  // even though the use of kernel9 happens in the false branch of a constexpr if
  // because both the true and the false branches cause the instantiation of kernel_single_task.
```

Reply:
I think this one may have been missed as well?

Reply:
Right. I was wondering about this one myself, and then realized that it was collapsed. Thank you.

Reply:
I think these changes are still missing.